scriptworker/artifacts.py

```diff
@@ -306,3 +306,39 @@ def get_expiration_arrow(context):
 
     await raise_future_exceptions(tasks)
     return files
+
+
+def get_upstream_artifacts(context):
+    """List the downloaded upstream artifacts.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+
+    Returns:
+        dict: the paths of uploaded artifacts, sorted per task_id
+
+    Raises:
+        scriptworker.exceptions.ScriptWorkerTaskException: when an artifact doesn't exist.
+    """
+    task_ids_and_relative_paths = [
+        (artifact_definition['taskId'], artifact_definition['paths'])
+        for artifact_definition in context.task['payload']['upstreamArtifacts']
+    ]
+
+    return {
+        task_id: [
+            _get_single_upstream_artifact(context, task_id, path)
+            for path in paths
+        ]
+        for task_id, paths in task_ids_and_relative_paths
+    }
+
+
+def _get_single_upstream_artifact(context, task_id, path):
```

Comment:
`_get_single_upstream_artifact_full_path` ?

Per our conversation in IRC:

- We also generate these full paths in `scriptworker.cot.verify`.  It would be good to reuse this function there, so we only have one place that defines these paths.
    - that may mean we drop the leading underscore
    - we probably want to avoid raising `ScriptWorkerTaskException` here, since the artifact won't exist on disk before we download.  We can either have a kwarg that says we should run this check, or run that check in `get_upstream_artifacts`.